### PR TITLE
PCI-1429 Update README file for Docker package manager

### PR DESCRIPTION
**Pull request by aliculPix4D**
- updating README file.

### CODE REVIEW

**Comment by marco-m-pix4d** [comment]
I am tempted to agree with @odormond that a separate file like README.PIX4D.md is better. Or is there a specific reason to add to the existing README ?

**Comment by Pix4D-Janus** [comment]
Your pipeline is set.
Check it here:https://builder.ci.pix4d.com/teams/developers/pipelines/dependabot-PCI-1429-dependabot-documentation

**Comment by aliculPix4D** [comment]
> I am tempted to agree with @odormond that a separate file like README.PIX4D.md is better. Or is there a specific reason to add to the existing README ?

I would say historic reasons, while importing the history from terraform-tomato (https://github.com/Pix4D/dependabot-core/commit/2df3f23ae68b7c187a9a68028f47e00b11a45266#diff-a489144702a723431811d56a3c44690a)
I am also in favor of renaming the file, as suggested.

**Comment by odormond** [comment]
What I'd like to see addressed in the README.PIX4D.md is what we're supposed to do when the checks fails as they are right now for this PR.

There is thousands of files failing the rubocop tests. I don't believe we're just supposed to reformat everything. How am I supposed to figure out what's happening? Does that mean some rebase failed? Is it the test that went boom due to some dependency changes between upstream and our test job? What are you going to do to make them pass?

**Comment by aliculPix4D** [comment]
> What I'd like to see addressed in the README.PIX4D.md is what we're supposed to do when the checks fails as they are right now for this PR.

For now, I have wrote this part here: https://docs.google.com/document/d/1iSHtWoRO5Yl2HF53ISanxDG1Qs0S8zwumWwLWH4x4mA/edit
it still needs review from the team. After which, I was planning to create a Confluence page for it. 
Here, I have just updated the old README.PIX4D.md file to be up-to-date (same explanation are in the document above).
I was not planning to put the full guide in the README.PIX4D.md. 
I propose the following: 
1) separate the README files 
2) README.PIX4D.md will contain the same text as it do now + the link to the Confluence page I will create

> There is thousands of files failing the rubocop tests. I don't believe we're just supposed to reformat everything. How am I supposed to figure out what's happening? Does that mean some rebase failed? Is it the test that went boom due to some dependency changes between upstream and our test job? What are you going to do to make them pass?

This was a special case, since all the tests were also failing upstream due to an issue with rubocop. This is now solved upstream by bumping the rubocop version: https://github.com/dependabot/dependabot-core/pull/2496
It is also already merged in our fork: https://github.com/Pix4D/dependabot-core/pull/58 so I will now rebase this PR on top of it.

Also, we never rebase, we merge the upstream changes. It is not yet happened that the merge job fail. For now it always succeded, failing tests then mean something is wrong and needs to be understood and fixed by adding an additional commit. Rubocop is also configured using upstream configuration file and since our files already comply to, if we did not modified our files, failing rubocop usually point to the error usptream.

**Comment by marco-m-pix4d** [comment]
@aliculPix4D  What is the reasoning behind using Confluence instead of keeping everything here ? I think that it is easier to find stuff in the same place (in this repo) as opposed to Confluence, but maybe I am missing something ? 
@odormond what would you prefer ?

**Comment by aliculPix4D** [comment]
> @aliculPix4D What is the reasoning behind using Confluence instead of keeping everything here ? I think that it is easier to find stuff in the same place (in this repo) as opposed to Confluence, but maybe I am missing something ?
I can update the README.PIX4D.md. I thought Confluence page is required, since we are adding other guides there, but maybe this is better and more easily accessible.

**Comment by odormond** [comment]
> This was a special case, since all the checks were also failing upstream due to an issue with rubocop. This is now solved upstream by bumping the rubocop version: [dependabot#2496](https://github.com/dependabot/dependabot-core/pull/2496)

Being special is one thing and being rare is another. As we had such a case when you were on vacation it seems like it might not be that rare. So I would suggest to details in the readme (preferably) how to detect it's a special case and how to handle it. Maybe something like:
```
If the checks fail, verify if the upstream checks (CI / CI (bundler), ...) fails (how do we do that?). If so, wait for an upstream fix. Otherwise, pull the branch and try to fix it.
```

> Also, we never rebase, we merge the upstream changes.

Right, sorry, my bad.

> Rubocop is also configured using upstream configuration file and since our files already comply to, if we did not modified our files, failing rubocop usually point to the error usptream.

That sounds like a good way to figure out if it broke upstream first or not. Now the question is: which files are ours? There is 700+ commits since our first commit. There might be a simple git way of finding all the files that were modified by one of use. If so, then it should be included in the readme because I don't think we do this kind of git trick very often and I certainly don't know how to do it.

Also the recent update of `black` showed that a perfectly fine piece of code can stop passing linter tests whenever the linter is updated.

**Comment by aliculPix4D** [comment]
> Being special is one thing and being rare is another. As we had such a case when you were on vacation it seems like it might not be that rare. So I would suggest to details in the readme (preferably) how to detect it's a special case and how to handle it. 

Now I think, I understand what is going on and what is confusing. In July (before holidays) there were only 1 check `test-dependabot`. It is required one, and it is a job in Concourse that runs test in docker folder (all of the tests, both Pix4D ones and the upstream ones) and the rubocop using upstream configuration file on the whole repository. 
In August they activated Github actions and now we have 18 new checks ( `CI / CI (docker)` should be almost eqvivalent to our tests). At least some of those checks were failing upstream until recently. What we are doing in the docker folder (docker packer manager), they repeat for each of the packer managers. also they run rubocop for each folder, while we run it for the whole repository. These checks are not required to pass, but from what I see, are a good indicator if rubocop will fail in our case. 

In short, I will update the document (README at the end) with the new findings and instructions.

**Comment by odormond** [comment]
> > Being special is one thing and being rare is another. As we had such a case when you were on vacation it seems like it might not be that rare. So I would suggest to details in the readme (preferably) how to detect it's a special case and how to handle it.
> 
> Now I think, I understand what is going on and what is confusing. In July (before holidays) there were only 1 check `test-dependabot`. It is required one, and it is a job in Concourse that runs test in docker folder (all of the tests, both Pix4D ones and the upstream ones) and the rubocop using upstream configuration file on the whole repository.
> In August they activated Github actions and now we have 18 new checks ( `CI / CI (docker)` should be almost eqvivalent to our tests). At least some of those checks were failing upstream until recently. What we are doing in the docker folder (docker packer manager), they repeat for each of the packer managers. also they run rubocop for each folder, while we run it for the whole repository. These checks are not required to pass, but from what I see, are a good indicator if rubocop will fail in our case.
> 
> In short, I will update the document (README at the end) with the new findings and instructions.

Ah, good. It start making sense now. :-)

Do you think it could be possible to "skip" upstream commits if they have these check failing? That would reduce the noise and we would attempt a merge only when upstream is clean.

File: docker/README.PIX4D.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+```
+
+To fix formatting bugs in a file use `-a` flag, e.g.
+
+```shell
+bundle exec rubocop -a <filename>
+```
+*NOTE*: The `-a` flag will apply the changes in place, i.e. it will override the file
+
+## CONGIGURATION
```
Comment (manuel-pix4d) [inline docker/README.PIX4D.md:147]: typo

File: docker/README.PIX4D.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+Some additional package managers that could be useful for the team: Ruby, Python, Go, Terraform
+
+### Dependabot workflow/components:
+
+File Fetcher - File Parser - Update Checker - File Updater - PR creator
+
+Periodically triggered Concourse job, it fetches the Dockerfile that is monitored (_File Fetcher_), parses the _FROM_ line in the Dockerfile (_File Parser_) to find the tag of the Docker image used. It further checks (_Update Checker_) if a newer tag (update) is available in Docker registry. If yes it updates the Docker file (_File Updater_) and opens a new pull request in the repository (_PR creator_). In short, it proposes the updates of base Docker images used in the Dockerfile.
+
+![Example of a PR opened by Dependabot in linux-image-build repository](dependabot_pr_example.png?raw=true "PR opened by Dependabot")
```
Comment (aliculPix4D) [inline docker/README.PIX4D.md:18]: For some reason it does not show the image in the Preview mode. All is fine locally. Any suggestion on what is wrong here?

File: docker/README.PIX4D.md [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+3. Test-dependabot job runs unit tests that contain both upstream tests and our Pix4D modifications. Additionally it also runs rubocop a Ruby static code analyzer and code formatter
+
+4. If tests are green and PR gets 2 approvals, it can be merged to the master branch
+
+5. Any change in docker packer manager folder or Pix4D Dockerfile triggers the build of new Docker image by linux-image-build-master pipeline
+
+_CASE:_ upstream changes are merged in the feature branch, PR is opened, but the tests (rspec or rubocop) fail. What should I do?
+_SOLUTION_ locally fetch the branch, fix the error, add a new commit, run the tests again. After the PR is approved and the test job is green again, merge the PR.
+
```
Comment (manuel-pix4d) [inline docker/README.PIX4D.md:82]: Ah, I saw the ci tasks. I understand now as you just fetch alpine image and then install what you need to test the dependabot. I guess it was not possible to create the full image before because it would have required a stable dependencies. Right ?

**Comment by aliculPix4D** [comment]
@marco-m-pix4d @odormond I think I have addressed all the comments now. Can you please review it again?
- created a new README.PIX4D.md
- copied the content from [Google Docs](https://docs.google.com/document/d/1iSHtWoRO5Yl2HF53ISanxDG1Qs0S8zwumWwLWH4x4mA/edit?ts=5f620747) to the new readme  
- explained the workflow with broken upstream checks
- list the files modified by PCI team:
```
git log --pretty="%H" --author="Pix4D" | while read commit_hash; do git show --oneline --name-only $commit_hash | tail -n+2; done | sort | uniq
```

File: docker/README.PIX4D.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+### SUPPORTED WORKFLOW
+
+1. merge-upstream job is automatically triggered
+    * fetches a new upstream version
+    * creates a feature branch in the forked repository
+    * merges upstream changes
+    * creates a PR in the forked repository
+       - PR title: [no-changes-to-pix4d-dependabot] - fast review process
+       - PR title: [changes-to-pix4d-dependabot] - more detailed review required
```
Comment (manuel-pix4d) [inline docker/README.PIX4D.md:68]: That is the problematic part. What do you review exactly ? As most of us don't know the logic of the tool, it is quite difficult to review (both actually, the `no-changes` and the `changes`). Could you explain a bit more what to check ? Otherwise, would it make sense to merge it automatically anytime the tests pass ?

**Review by manuel-pix4d — COMMENTED** [review]
Thanks for the documentation, it's a bit clearer what we should do. But I'm still confused about what to review in these PRs.

File: docker/README.PIX4D.md [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+2. The depenabot-core repository is monitored by set-pipeline job in Github-automation-tools pipeline - since the new PR is opened by the merge-upstream job,, PR resource detects it, sets a new featured pipeline, writes a comment in the PR with the pipeline URL and triggers the test-dependabot job
+
+3. Test-dependabot job runs unit tests that contain both upstream tests and our Pix4D modifications. Additionally it also runs rubocop a Ruby static code analyzer and code formatter
+
+4. If tests are green (passing test-dependabot is required) and PR gets 2 approvals, it can be merged to the master branch
+
+5. Any change in docker packer manager folder or Pix4D Dockerfile, that is merged to master branch, triggers the build of new Docker image by linux-image-build-master pipeline
```
Comment (odormond) [inline docker/README.PIX4D.md:76]: Typo:
```suggestion
5. Any change in docker package manager folder or Pix4D Dockerfile, that is merged to master branch, triggers the build of new Docker image by linux-image-build-master pipeline
```

**Review by odormond — APPROVED** [review]
Thanks a lot for this PR. It will be highly valuable to help fix things when broken.

File: docker/README.PIX4D.md (hunk 3 above)
Comment (aliculPix4D) [reply docker/README.PIX4D.md:82]: Yes, you are correct. Dependencies versions are changed upstream by gemspec file in [common folder](https://github.com/Pix4D/dependabot-core/blob/master/common/dependabot-common.gemspec) and each time the job is triggered a new version of gem common is created (with its own dependencies), hence we need to install correct dependencies each time.

File: docker/README.PIX4D.md (hunk 2 above)
Comment (marco-m-pix4d) [reply docker/README.PIX4D.md:18]: if you look on the branch page, do you see it ?
https://github.com/Pix4D/dependabot-core/blob/PCI-1429-dependabot-documentation/docker/README.PIX4D.md

File: docker/README.PIX4D.md (hunk 4 above)
Comment (aliculPix4D) [reply docker/README.PIX4D.md:68]: For the: `PR title: [no-changes-to-pix4d-dependabot] - fast review process`, I would agree maybe, there is not a lot we can do here and review (maybe just look a little on what they changed in the common directory and overall rubocop changes). We can merge the PR if is green, or not merge it if job fails. We never build the new Docker image, based on the changes made on this PR. Hence, yes it mostly creates noise and confusion.  Auto merging this kind of PRs would probably be the best option, if the team agrees. It would also solve @odormond question which I forgot to answer previously:

> Do you think it could be possible to "skip" upstream commits if they have these check failing? That would reduce the noise and we would attempt a merge only when upstream is clean.

For the second PR:
`PR title: [changes-to-pix4d-dependabot] - more detailed review required`
From March 2020 until now, there were exactly 2 PR of this type:
https://github.com/Pix4D/dependabot-core/pull/32
https://github.com/Pix4D/dependabot-core/pull/40

I agree that they are confusing, because they contain a lot of files. But we should review only files in docker folder by filtering them in Github GUI. From here we can immediately see that in the second PR only 4 fixtures files were changed and the only change is string change from `master` ----> `main` (https://github.com/Pix4D/dependabot-core/pull/40/files#diff-2a2057abf949a11de8c953661163bf17).
In the first PR 2 files changed, one of each is [update checker](https://github.com/Pix4D/dependabot-core/pull/32/files#diff-29bed84182b93f082c3ab4aa29ba02e3). In this case it is imperative that all rspec test pass and reviewing the changes is required in my opinion.

My proposal for this PR:
- for the PR titled [no-changes-to-pix4d-dependabot] I change fast review process to no review needed if tests are passing, otherwise reject the PR with explanation
-  I add only a few lines with explanation about how to review the PR title with title:  `[changes-to-pix4d-
dependabot] - more detailed review required`
- with team agreement, in a separate PR I change the merge-upstream job, to auto merge the PR with title: [no-changes-to-pix4d-dependabot]

File: docker/README.PIX4D.md (hunk 2 above)
Comment (aliculPix4D) [reply docker/README.PIX4D.md:18]: Yes. Thanks.

**Review by marco-m-pix4d — APPROVED** [review]

File: docker/README.PIX4D.md (hunk 4 above)
Comment (manuel-pix4d) [reply docker/README.PIX4D.md]: It would be perfect ! Thanks Aleksander !

**Comment by manuel-pix4d** [comment]
Just hit me now. Would it be better to put this PIX4D readme inside the root folder, instead of docker/ ? Just to make it easier to get once you search for it.

**Comment by aliculPix4D** [comment]
> Just hit me now. Would it be better to put this PIX4D readme inside the root folder, instead of docker/ ? Just to make it easier to get once you search for it.

I have added a line in docker README pointing to [Pix4D readme](https://github.com/Pix4D/dependabot-core/tree/master/docker#pix4d-dependabot-readme-file)
